src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs

```diff
@@ -2172,8 +2172,10 @@ private void ReportMethodSignatureCollision(BindingDiagnosticBag diagnostics, Me
                 return;
             }
 
-            if ((method1 as SourceExtensionImplementationMethodSymbol)?.UnderlyingMethod.ContainingType is { } extensionDeclaration &&
-                (object)extensionDeclaration == (method2 as SourceExtensionImplementationMethodSymbol)?.UnderlyingMethod.ContainingType)
+            if (method1 is SourceExtensionImplementationMethodSymbol { UnderlyingMethod: var underlying1, UnderlyingMethod.ContainingType: { } extensionDeclaration1 } &&
+                method2 is SourceExtensionImplementationMethodSymbol { UnderlyingMethod: var underlying2, UnderlyingMethod.ContainingType: { } extensionDeclaration2 } &&
+                (object)extensionDeclaration1 == extensionDeclaration2 &&
```

Comment:
nit: we probably don't need to name the containing types. `(object)underlying1.UnderlyingMethod.ContainingType == underlying2.UnderlyingMethod.ContainingType` may be less cluttered